File: src/output.ts
import { type Envelope, SCHEMA_VERSION } from "./types";

export const okEnvelope = <T>(command: string, data: T): Envelope<T> => ({
  schema_version: SCHEMA_VERSION,
  command,
  ok: true,
  data,
});

export const errEnvelope = (
  command: string,
  code: string,
  message: string,
  details?: unknown,
): Envelope<never> => ({
  schema_version: SCHEMA_VERSION,
  command,
  ok: false,
  error: {
    code,
    message,
    ...(details === undefined ? {} : { details }),
  },
});

File: src-rust/src/output.rs
use crate::types::{Envelope, EnvelopeErr, EnvelopeError, EnvelopeOk, SCHEMA_VERSION};
use serde_json::Value;

pub fn ok_envelope<T>(command: impl Into<String>, data: T) -> Envelope<T> {
    Envelope::Ok(EnvelopeOk {
        schema_version: SCHEMA_VERSION,
        command: command.into(),
        ok: true,
        data,
    })
}

pub fn err_envelope(
    command: impl Into<String>,
    code: impl Into<String>,
    message: impl Into<String>,
    details: Option<Value>,
) -> Envelope<Value> {
    Envelope::Err(EnvelopeErr {
        schema_version: SCHEMA_VERSION,
        command: command.into(),
        ok: false,
        error: EnvelopeError {
            code: code.into(),
            message: message.into(),
            details,
        },
    })
}
